CppGraphFiles/NumberOfShortestPathsMatrix.cpp

```diff
@@ -0,0 +1,522 @@
+                ans += toPush;
+                if (j == other.number.length() - 1 && carry != 0) 
+                {
+                    char extraPush = carry + '0';
+                    ans += extraPush;
+                }
+            }
+            for (int k = 0; k < this->number.length() - i - 1; ++k) 
+            {
+                ans = '0' + ans;
+            }
+            std::reverse(ans.begin(), ans.end());
+            BigInt add(ans);
+            result.push_back(add);
+        }
+        BigInt sum("0");
+        for (int i = 0; i < result.size(); ++i) 
+        {
+            sum = sum + result[i];
+        }
+        return sum;
+    }
+
+    friend std::ostream& operator<<(std::ostream& out, const BigInt& bin)
+    {
+        out << bin.number;
+        out << std::endl;
+        return out;
+    }
+
+    void operator++(int)
+    {
+        BigInt newNumber(this->number);
+        newNumber = newNumber + BigInt("1");
+        this->number = newNumber.number;
+    }
+
+    void operator++()
+    {
+        BigInt newNumber(this->number);
+        newNumber = newNumber + BigInt("1");
+        this->number = newNumber.number;
+    }
+};
+
+BigInt integerToBigInt(int n)
+{
+    std::string result = "";
+    while (n > 0)
+    {
+        int digit = n % 10;
+        n /= 10;
+        char toPush = digit + '0';
+        result += toPush;
+    }
+    std::reverse(result.begin(), result.end());
+    BigInt temp(result);
+    return temp;
+}
+
+int min(int a, int b)
+{
+    return (a < b) ? a : b;
+}
+
+BigInt BinomialCoeff(int n, int k)
+{
+    std::vector<std::vector<BigInt>> C(n + 1, std::vector<BigInt>(k + 1));
+    int i, j;
+
+    for (i = 0; i <= n; ++i)
+    {
+        for (j = 0; j <= min(i, k); ++j)
+        {
+            if (j == 0 || j == i)
+            {
+                C[i][j] = BigInt("1");
+            }
+            else
+            {
+                C[i][j] = C[i - 1][j - 1] + C[i - 1][j];
+            }
+        }
+    }
+    return C[n][k];
+}
+
+bool VertexInRnge(int vertex, int verticesCount)
+{
+    return vertex >= 0 && vertex < verticesCount;
+}
+
+adjacency_list ReadGraph(std::ifstream& in)
+{
+    int vertices, edges;
+
+    in >> vertices >> edges;
+
+    if (vertices < 0 || edges < 0)
+    {
+        throw std::logic_error("Wrong value for vertices or edges");
+    }
+
+    adjacency_list graph(vertices);
+
+    for (int i = 0; i < edges; ++i)
+    {
+        int vertex1, vertex2, weight;
+
+        in >> vertex1 >> vertex2 >> weight;
+
+        if (VertexInRnge(vertex1, vertices) && VertexInRnge(vertex2, vertices) && weight >= 0)
+        {
+            graph[vertex1].push_back(std::make_pair(vertex2, weight));
+            graph[vertex2].push_back(std::make_pair(vertex1, weight));
+        }
+        else
+        {
+            throw std::logic_error("Wrong edge or wrong weight");
+        }
+    }
+
+    return graph;
+}
+
+Coordinate FindCoordinates(int vertex)
+{
+    int y = vertex % 100;
```

Comment:
what is going on here???

Reply:
This function founds the coordinate of the vertex in a matrix graph. It is used to calculate the number of all shortest paths using binomial coefficient